enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,348 @@
+  - `timeWindow`: the time CPU and memory usage have to be below the thresholds,
+    before triggering a scale down event.
+
+A complete configuration in YAML format could then look like this:
+
+```yaml
+controlplaneautoscaling:
+  instanceSizePath: spec.template.spec.providerSpec.value.instanceType
+  instanceSizes:
+  - weight: 1
+    name: r4.large
+  - weight: 2
+    name: r4.xlarge
+  - weight: 4
+    name: r4.2xlarge
+  # [mandatory configuration]
+  syncPeriod: [duration as specified in golang: e.g. 5s, 15s]
+  gracePeriod: [duration as specified in golang: e.g. 5s, 15s]
+  scaledownEnabled: [true|false]
+  scaleUpThresholds:
+    cpuLoadAverage: [0-100 as percentage]
+    memoryUsage: [0-100 as percentage]
+    timeWindow: [duration as specified in golang: e.g. 5m, 15m]
+  scaleDownThresholds:
+    cpuLoadAverage: [0-100 as percentage]
+    memoryUsage: [0-100 as percentage]
+    timeWindow: [duration as specified in golang: e.g. 5m, 15m]
+```
+
+### Data gathering
+
+As specified the operator should not require access to data from operators that
+might not be installed in the cluster (e.g. `prometheus`).
+
+Instead the operator should gather the required data via the [metrics
+API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/)
+that is also used for the horizontal and vertical pod autoscalers.
+
+Data available from the metrics API includes:
+
+- CPU usage in `cores`.
+- Memory usage in a SI units.
+- The time window used to aggregate the above mentioned data.
+
+To gather data about the available CPU cores and memory available to control
+plane nodes, the operator should use the kubernetes
+[Node](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#node-v1-core)
+API.
+
+As the window returned by metrics API is likely shorter than required window for
+scaling decisions, the operator should gather its own average to allow making
+the scaling decision based on the metrics data.
+
+### Workflow for automatic scaling
+
+Automatic scaling should be triggered the same as [horizontal pod
+autoscaling](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) -
+that runs periodically to check if scaling the control-plane machines is
+required.
+
+In addition the implementation has to ensure that scaling decision are not based
+on single point-in-time values to prevent scaling up or down based on usage
+spikes.
+
+Instead a running average (or similar) should be used to decide about scaling.
+
+If for any reason metrics can not be gathered, the operator must remain
+available and only the autoscaling should be disabled / non-functional.
+
+### Additional RBAC permissions
+
+The operator will require access to the following resources to gather the data:
+
+- Read access for `/apis/metrics.k8s.io/v1beta1/nodes`.
+- Read access for `/api/v1/nodes` (this seems already in place for OpenShift,
+  but has to be verified for a non-OpenShift installations).
```

Comment:
Non openshift?

Reply:
I am not sure how we handle vanilla Kubernetes clusters in this proposals, so I included that point in this list.
If vanilla Kubernetes installations can be ignored, I can just remove this one.
However in that case it might be better to say `non-OSD` installations, as that is the one I checked against.

Reply:
Yeah we don't tend to worry about non-openshift. There's a difference between openshift, OSD, HyperShift, MicroShift to account for but, we can assume everything this proposal will apply to will be some kind of openshift